### PR TITLE
feat: Backport first part of improved theming support

### PR DESCRIPTION
**Pull request by caalador**
Backport of initial parts of the improved theming support.

Part of #9983

### CODE REVIEW

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/FullDependenciesScannerTest.java [hunk 1]
```diff
@@ -73,7 +73,7 @@ public String getThemeUrl() {
 
     }
 
-    @Theme(FakeLumoTheme.class)
+    @Theme(value =FakeLumoTheme.class)
```
Comment (taefi) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/FullDependenciesScannerTest.java:76]:
```suggestion
    @Theme(value = FakeLumoTheme.class)
```

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/ScannerTestComponents.java [hunk 2]
```diff
@@ -108,7 +108,7 @@ public String getThemeUrl() {
     }
 
 
-    @Theme(value = Theme1.class, variant = Theme0.DARK)
+    @Theme(value =Theme1.class, variant = Theme0.DARK)
```
Comment (taefi) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/ScannerTestComponents.java:111]:
```suggestion
    @Theme(value = Theme1.class, variant = Theme0.DARK)
```

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/ScannerTestComponents.java [hunk 3]
```diff
@@ -117,7 +117,7 @@ public Element getElement() {
         }
     }
 
-    @Theme(value = Theme1.class, variant = Theme0.DARK)
+    @Theme(value =Theme1.class, variant = Theme0.DARK)
```
Comment (taefi) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/ScannerTestComponents.java:120]:
```suggestion
    @Theme(value = Theme1.class, variant = Theme0.DARK)
```

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/ScannerTestComponents.java [hunk 4]
```diff
@@ -127,7 +127,7 @@ public Element getElement() {
     }
 
     @Route(value = "")
-    @Theme(Theme4.class)
+    @Theme(value =Theme4.class)
```
Comment (taefi) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/ScannerTestComponents.java:130]:
```suggestion
    @Theme(value = Theme4.class)
```

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/ScannerTestComponents.java [hunk 5]
```diff
@@ -151,7 +151,7 @@ public Element getElement() {
     }
 
     @Route(value = "", layout = RouterLayout2.class)
-    @Theme(value = Theme2.class, variant = Theme2.FOO)
+    @Theme(value =Theme2.class, variant = Theme2.FOO)
```
Comment (taefi) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/ScannerTestComponents.java:154]:
```suggestion
    @Theme(value = Theme2.class, variant = Theme2.FOO)
```

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/ScannerTestComponents.java [hunk 6]
```diff
@@ -233,7 +233,7 @@ public void configureInstance(WebComponent<RootViewWithTheme> webComponent, Root
         }
     }
 
-    @Theme(Theme2.class)
+    @Theme(value =Theme2.class)
```
Comment (taefi) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/ScannerTestComponents.java:236]:
```suggestion
    @Theme(value = Theme2.class)
```

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/samples/RouteLayoutComponent.java [hunk 7]
```diff
@@ -22,7 +22,7 @@
 import com.vaadin.flow.theme.Theme;
 
 @JsModule("foo.js")
-@Theme(CustomTheme.class)
+@Theme(value =CustomTheme.class)
```
Comment (taefi) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/scanner/samples/RouteLayoutComponent.java:25]:
```suggestion
@Theme(value = CustomTheme.class)
```

File: flow-tests/test-themes/frontend/theme/app-theme/global.css [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+@font-face {
+    font-family: "Ostrich";
+    src: url("./font/ostrich-sans-regular.ttf") format("TrueType");
+}
+
+body {
+    background-image: url("./img/bg.jpg");
+    font-family: "Ostrich";
+}
```
Comment (taefi) [inline flow-tests/test-themes/frontend/theme/app-theme/global.css:11]: Just to make sure:
This file misses the following lines from #9349 
```
body {
    background-image: url('theme/app-theme/img/bg.jpg');
}
```

**Review by taefi — CHANGES_REQUESTED** [review]

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java [hunk 9]
```diff
@@ -296,21 +298,39 @@ private void computeApplicationTheme() throws ClassNotFoundException,
...
         if (themes.isEmpty()) {
             theme = getDefaultTheme();
         } else {
             // we have a proper theme or no-theme for the app
             ThemeData themeData = themes.iterator().next();
             if (!themeData.isNotheme()) {
+                String themeClass = themeData.getThemeClass();
+                if (!themeData.getThemeName().isEmpty() && (themeClass != null
+                    && getDefaultTheme() != null && !getDefaultTheme()
+                    .isAssignableFrom(getFinder().loadClass(themeClass)))) {
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java:311]: It's part of the 14 requirements and is added content

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FullDependenciesScanner.java [hunk 10]
```diff
@@ -308,7 +309,8 @@ private ThemeData verifyTheme() {
                     .map(theme -> new ThemeData(
                             ((Class<?>) invokeAnnotationMethod(theme, VALUE))
                                     .getName(),
-                            invokeAnnotationMethodAsString(theme, "variant")))
+                            invokeAnnotationMethodAsString(theme, "variant"),
+                            invokeAnnotationMethodAsString(theme, "themeFolder")))
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FullDependenciesScanner.java:313]: It's 14 requirement as 14 doesn't use themeClass but value. Also themeFolder is value in 19

File: flow-server/src/main/java/com/vaadin/flow/theme/Theme.java [hunk 11]
```diff
@@ -29,12 +29,27 @@
...
+ * the theme is present inside the project, it maps to the {@code
+ * frontend/theme/<application-theme-name>}/ folder or alternatively to a folder
+ * inside the static resources of a jar file, like {@code
+ * src/main/resources/META-INF/resources/theme/<application-theme-name>/}.
+ * The application theme is always based on Lumo theme and this is the
+ * recommended way to theme applications starting from Flow 2.6 and Vaadin 14.6
+ * <p>
+ * Alternatively , if instead of Lumo theme the Material theme or another "old
+ * style custom theme" is to be used, that can be specified with the {@code
+ * value} property.
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/theme/Theme.java:42]: This is the requirement for 14 backport.

File: flow-server/src/main/java/com/vaadin/flow/theme/ThemeDefinition.java [hunk 12]
```diff
@@ -60,7 +65,7 @@ public ThemeDefinition(Class<? extends AbstractTheme> theme,
      *            the annotation to get the definition from
      */
     public ThemeDefinition(Theme themeAnnotation) {
-        this(themeAnnotation.value(), themeAnnotation.variant());
+        this(themeAnnotation.value(), themeAnnotation.variant(), themeAnnotation.themeFolder());
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/theme/ThemeDefinition.java:68]: 14 backport requirement for the annotation values

File: flow-tests/test-themes/frontend/theme/app-theme/global.css (hunk 8 above)
Comment (caalador) [reply flow-tests/test-themes/frontend/theme/app-theme/global.css:11]: This is the correct version at this point in time see 9451

**Review by caalador — COMMENTED** [review]
As I'm out the whole week you can fix the formatting problems, also check the backport ticket for the Theme annotation requirements.

File: flow-maven-plugin/src/test/java/com/vaadin/flow/plugin/maven/TestComponents.java [hunk 13]
```diff
@@ -82,7 +82,7 @@
     }
 
     @Route
-    @Theme(value = Lumo.class, variant = Lumo.DARK)
+    @Theme(value =Lumo.class, variant = Lumo.DARK)
```
Comment (taefi) [inline flow-maven-plugin/src/test/java/com/vaadin/flow/plugin/maven/TestComponents.java:85]:
```suggestion
    @Theme(value = Lumo.class, variant = Lumo.DARK)
```

**Review by taefi — CHANGES_REQUESTED** [review]

**Review by taefi — DISMISSED** [review]

**Review by taefi — APPROVED** [review]

**Pull request merged**